tests/callbacks_test.go

```diff
@@ -169,100 +171,146 @@ func (s *FunctionalSuite) TestWorkflowCallbacks_InvalidArgument() {
+				return 666, nil
+			},
+			runTimeout: 100 * time.Millisecond,
+		},
+		{
+			name: "WorkflowFailureRetry",
+			wf: func(ctx workflow.Context) (int, error) {
+				// Verify that the callback is carried over the retry on failure boundary.
+				info := workflow.GetInfo(ctx)
+				if info.FirstRunID == info.WorkflowExecution.RunID {
+					return 0, errors.New("intentional workflow failure")
+				}
+				s.Greater(info.Attempt, int32(1))
+				return 666, nil
+			},
+			runTimeout: 100 * time.Second,
+		},
+	}
 
-	taskQueue := s.randomizeStr(s.T().Name())
-	workflowType := "test"
+	for _, tc := range cases {
+		s.T().Run(tc.name, func(t *testing.T) {
+			ctx := NewContext()
+			sdkClient, err := client.Dial(client.Options{
+				HostPort:  s.testCluster.GetHost().FrontendGRPCAddress(),
+				Namespace: s.namespace,
+			})
+			s.NoError(err)
+			pp := temporalite.NewPortProvider()
 
-	w := worker.New(sdkClient, taskQueue, worker.Options{})
-	wf := func(ctx workflow.Context) (int, error) {
-		// Verify that the callback is carried over the CAN boundary.
-		if workflow.GetInfo(ctx).ContinuedExecutionRunID == "" {
-			return 0, workflow.NewContinueAsNewError(ctx, workflowType)
-		}
-		return 666, nil
-	}
-	ch := &completionHandler{
-		requestCh:         make(chan *nexus.CompletionRequest, 1),
-		requestCompleteCh: make(chan error, 1),
-	}
-	callbackAddress := fmt.Sprintf("localhost:%d", pp.MustGetFreePort())
-	s.NoError(pp.Close())
-	shutdownServer := s.runNexusCompletionHTTPServer(ch, callbackAddress)
-	defer func() {
-		err := shutdownServer()
-		if err != nil {
-			panic(err)
-		}
-	}()
-	w.RegisterWorkflowWithOptions(wf, workflow.RegisterOptions{Name: workflowType})
-	s.NoError(w.Start())
-	defer w.Stop()
+			taskQueue := s.randomizeStr(s.T().Name())
+			workflowType := "test"
 
-	// TODO: use sdkClient instead of directly calling the history engine when callbacks are exposed
-	request := &workflowservice.StartWorkflowExecutionRequest{
-		RequestId:          uuid.New(),
-		Namespace:          s.namespace,
-		WorkflowId:         s.randomizeStr(s.T().Name()),
-		WorkflowType:       &commonpb.WorkflowType{Name: workflowType},
-		TaskQueue:          &taskqueuepb.TaskQueue{Name: taskQueue, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
-		Input:              nil,
-		WorkflowRunTimeout: durationpb.New(100 * time.Second),
-		Identity:           s.T().Name(),
-		CompletionCallbacks: []*commonpb.Callback{
-			{
-				Variant: &commonpb.Callback_Nexus_{
-					Nexus: &commonpb.Callback_Nexus{
-						Url: "http://" + callbackAddress,
+			w := worker.New(sdkClient, taskQueue, worker.Options{})
+
+			ch := &completionHandler{
+				requestCh:         make(chan *nexus.CompletionRequest, 1),
+				requestCompleteCh: make(chan error, 1),
+			}
+			callbackAddress := fmt.Sprintf("localhost:%d", pp.MustGetFreePort())
+			s.NoError(pp.Close())
+			shutdownServer := s.runNexusCompletionHTTPServer(ch, callbackAddress)
+			defer func() {
+				require.NoError(t, shutdownServer())
```

Comment:
I think you can't use `require` in a defer statement but I'm not 100% sure. You could also use `t.Cleanup` instead.